File: day5.py
#this is extremely slow do not run

with open("inputs/day5.txt", "r") as f:
    input = f.read()

lines = input.splitlines()[2:-1]

seeds = [int(num) for num in input.splitlines()[0].split(": ")[1].split()]
maps = {}
currentMap = ""

for line in lines:
    if "map" in line:
        currentMap = line.split(":")[0].split()[0]
        maps[currentMap] = []
    elif line != "":
        currentDictionary = {}
        currentDictionary["drstart"] = int(line.split()[0])
        currentDictionary["srstart"] = int(line.split()[1])
        currentDictionary["rlength"] = int(line.split()[2])
        maps[currentMap].append(currentDictionary)

mapsComplete = {}

for key, value in maps.items():
    mapsComplete[key] = {}
    for dictionaryno in range(len(value)):
        destinationRange = range(value[dictionaryno]["drstart"], value[dictionaryno]["drstart"]+value[dictionaryno]["rlength"])
        sourceRange = range(value[dictionaryno]["srstart"], value[dictionaryno]["srstart"]+value[dictionaryno]["rlength"])
        if mapsComplete[key] == {}:
            mapsComplete[key] = {"destinationRange": list(destinationRange), "sourceRange": list(sourceRange)}
        else:
            mapsComplete[key]["destinationRange"] += destinationRange
            mapsComplete[key]["sourceRange"] += list(sourceRange)

locationNums = []

for seed in seeds:
    currentNum = seed
    for key, map in mapsComplete.items():
        if currentNum in map["sourceRange"]:
            numIndex = map["sourceRange"].index(currentNum)
            currentNum = map["destinationRange"][numIndex]
        else:
            currentNum = currentNum
    locationNums.append(currentNum)

print(min(locationNums))

File: day5.mjs
import { readFile } from "node:fs/promises";

const input = await readFile("inputs/day5.txt", "utf8");
const allLines = input.split("\n");
const firstLine = allLines[0];
// Line 2 is an empty line
const lines = allLines.slice(2);

const seedsText = firstLine.split(": ")[1];
const seedTexts = seedsText.split(" ");
const seeds = seedTexts.map((seedText) => parseInt(seedText, 10));

let maps = {};
let currentMap = "";

for (let i = 0; i < lines.length; i++) {
  const line = lines[i];
  const s = line.split(" ");
  if (line.includes("map")) {
    currentMap = s[0];
    maps[currentMap] = { destinationrange: [], sourcerange: [] };
  } else if (line != "") {
    let currentObject = maps[currentMap];
    let drstart = parseInt(s[0], 10);
    let srstart = parseInt(s[1], 10);
    let rangelength = parseInt(s[2], 10);
    currentObject["destinationrange"].push([drstart, drstart + rangelength]);
    currentObject["sourcerange"].push([srstart, srstart + rangelength]);
  }
}

let locations = [];

for (let seedIndex = 0; seedIndex < seeds.length; seedIndex++) {
  let seed = seeds[seedIndex];
  let seedToEnd = seed;
  let containerNumsIn;
  for (const [key, { sourcerange, destinationrange }] of Object.entries(maps)) {
    let numInSource = false;
    for (let contId = 0; contId < sourcerange.length; contId++) {
      let container = sourcerange[contId];
      if (seedToEnd > container[0] && seedToEnd < container[1]) {
        numInSource = true;
        containerNumsIn = container;
      }
    }
    if (numInSource) {
      let difference = sourcerange.indexOf(containerNumsIn);
      let numIndex = seedToEnd - containerNumsIn[0];
      seedToEnd = destinationrange[difference][0] + numIndex;
    } else {
      seedToEnd = seedToEnd;
    }
  }
  locations.push(seedToEnd);
}

console.log(Math.min(...locations));
